pkg/adapter/kubernetesevents/adapter.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetesevents
+
+import (
+	"context"
+	"github.com/knative/pkg/logging"
+	"github.com/knative/pkg/signals"
+	"go.uber.org/zap"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/knative/pkg/cloudevents"
+	corev1 "k8s.io/api/core/v1"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	eventType = "dev.knative.k8s.event"
+)
+
+type Adapter struct {
+	// Namespace is the kubernetes namespace to watch for events from.
+	Namespace string
+	// SinkURI is the URI messages will be forwarded on to.
+	SinkURI string
+
+	ctx context.Context
+}
+
+// Run starts the kubernetes receive adapter and sends events produced in the
+// configured namespace and send them to the sink uri provided.
+func (a *Adapter) Run(ctx context.Context) {
+	a.ctx = ctx
+	logger := logging.FromContext(ctx)
+
+	// set up signals so we handle the first shutdown signal gracefully
+	stopCh := signals.SetupSignalHandler()
+
+	cfg, err := clientcmd.BuildConfigFromFlags("", "")
+	if err != nil {
+		logger.Fatalf("error building kubeconfig: %v", zap.Error(err))
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		logger.Fatalf("error building kubernetes clientset: %v", zap.Error(err))
+	}
+
+	eventsInformer := coreinformers.NewFilteredEventInformer(
+		kubeClient, a.Namespace, 0, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}, nil)
+
+	eventsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    a.addEvent,
+		UpdateFunc: a.updateEvent,
+	})
```

Comment:
Optional, but this looks a lot like a controller. I wonder if we could save some lines (and maybe some bugs) by making this a standard controller-runtime reconciler. A subtlety is that we'd have to set the Manager's `Resync` option to `0` to disable resync so we don't get duplicate sends.

Reply:
I think this is a great idea. But perhaps it is a refactor after?